lib/widgets/emoji_reaction.dart

```diff
@@ -360,3 +366,261 @@ class _TextEmoji extends StatelessWidget {
+    if (!context.mounted) return;
+
+    switch (e) {
+      case ZulipApiException():
+        errorMessage = e.message;
+        // TODO(#741) specific messages for common errors, like network errors
+        //   (support with reusable code)
+      default:
+        // TODO(log)
+    }
+
+    showErrorDialog(context: context,
+      title: errorDialogTitle,
+      message: errorMessage);
+    return;
+  }
+}
+
+/// Opens a browsable and searchable emoji picker bottom sheet.
+void showEmojiPickerSheet({
+  required BuildContext pageContext,
+  required Message message,
+}) {
+  final store = PerAccountStoreWidget.of(pageContext);
+  showModalBottomSheet<void>(
+    context: pageContext,
+    // Clip.hardEdge looks bad; Clip.antiAliasWithSaveLayer looks pixel-perfect
+    // on my iPhone 13 Pro but is marked as "much slower":
+    //   https://api.flutter.dev/flutter/dart-ui/Clip.html
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    builder: (BuildContext context) {
+      return SafeArea(
+        child: Padding(
+          // By default, when software keyboard is opened, the ListView
+          // expands behind the software keyboard — resulting in some
+          // list entries being covered by the keyboard. Add explicit
+          // bottom padding the size of the keyboard, which fixes this.
+          padding: EdgeInsets.only(bottom: MediaQuery.viewInsetsOf(context).bottom),
+          // For _EmojiPickerItem, and RealmContentNetworkImage used in ImageEmojiWidget.
+          child: PerAccountStoreWidget(
+            accountId: store.accountId,
+            child: EmojiPicker(pageContext: pageContext, message: message))));
+    });
+}
+
+@visibleForTesting
+class EmojiPicker extends StatefulWidget {
+  const EmojiPicker({
+    super.key,
+    required this.pageContext,
+    required this.message,
+  });
+
+  final BuildContext pageContext;
+  final Message message;
+
+  @override
+  State<EmojiPicker> createState() => _EmojiPickerState();
+}
+
+class _EmojiPickerState extends State<EmojiPicker> with PerAccountStoreAwareStateMixin<EmojiPicker> {
+  late TextEditingController _controller;
+
+  EmojiAutocompleteView? _viewModel;
+  List<EmojiAutocompleteResult> _resultsToDisplay = const [];
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = TextEditingController()
+      ..addListener(_handleControllerUpdate);
+  }
+
+  void _initViewModel(EmojiAutocompleteQuery query) {
+    final store = PerAccountStoreWidget.of(context);
+    _viewModel = EmojiAutocompleteView.init(store: store, query: query)
+      ..addListener(_handleViewModelUpdate);
+  }
+
+  @override
+  void onNewStore() {
+    if (_viewModel != null) {
+      _viewModel!.dispose();
+    }
+    _initViewModel(EmojiAutocompleteQuery(_controller.text));
+  }
+
+  void _handleControllerUpdate() {
+    if (_viewModel != null) {
+      _viewModel!.query = EmojiAutocompleteQuery(_controller.text);
+    }
```

Comment:
It should be impossible for `_viewModel` to be null at this point, right? (Similar reasoning to at https://github.com/zulip/zulip-flutter/pull/1103#discussion_r1874217404 .) So:
```suggestion
  void _handleControllerUpdate() {
    _viewModel!.query = EmojiAutocompleteQuery(_controller.text);
```